File: csharp/ContextGate.Core/Services/ConfigManager.cs
using ContextGate.Core.Models;
using ContextGate.Core.Services.Interfaces;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace ContextGate.Core.Services;

/// <summary>
/// 配置管理服务实现
/// </summary>
public class ConfigManager : IConfigManager
{
    private readonly ISerializer _serializer;
    private readonly IDeserializer _deserializer;
    private AppConfig? _cachedConfig;

    public string ConfigPath { get; }

    public ConfigManager(string configPath)
    {
        ConfigPath = configPath;
        
        _serializer = new SerializerBuilder()
            .WithNamingConvention(UnderscoredNamingConvention.Instance)
            .Build();
            
        _deserializer = new DeserializerBuilder()
            .WithNamingConvention(UnderscoredNamingConvention.Instance)
            .IgnoreUnmatchedProperties()
            .Build();
    }

    public AppConfig LoadConfig()
    {
        if (_cachedConfig != null)
            return _cachedConfig;

        if (!File.Exists(ConfigPath))
        {
            // 创建默认配置
            _cachedConfig = CreateDefaultConfig();
            SaveConfig(_cachedConfig);
            return _cachedConfig;
        }

        try
        {
            var yaml = File.ReadAllText(ConfigPath);
            _cachedConfig = _deserializer.Deserialize<AppConfig>(yaml) ?? CreateDefaultConfig();
            return _cachedConfig;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Failed to load config: {ex.Message}");
            _cachedConfig = CreateDefaultConfig();
            return _cachedConfig;
        }
    }

    public bool SaveConfig(AppConfig config)
    {
        try
        {
            var yaml = _serializer.Serialize(config);
            
            // 确保目录存在
            var directory = Path.GetDirectoryName(ConfigPath);
            if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
            {
                Directory.CreateDirectory(directory);
            }
            
            File.WriteAllText(ConfigPath, yaml);
            _cachedConfig = config;
            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Failed to save config: {ex.Message}");
            return false;
        }
    }

    public ProviderConfig? GetProvider(string name)
    {
        var config = LoadConfig();
        return config.Providers.TryGetValue(name, out var provider) ? provider : null;
    }

    public void SetWorkspace(string path)
    {
        var config = LoadConfig();
        config.Workspace = path;
        SaveConfig(config);
    }

    public string? GetWorkspace()
    {
        var config = LoadConfig();
        return config.Workspace;
    }

    public Dictionary<string, ProviderConfig> GetAllProviders()
    {
        var config = LoadConfig();
        return config.Providers;
    }

    public ProxyConfig GetProxyConfig()
    {
        var config = LoadConfig();
        return config.Proxy;
    }

    public MonitorConfig GetMonitorConfig()
    {
        var config = LoadConfig();
        return config.Monitor;
    }

    public CurrencyConfig GetCurrencyConfig()
    {
        var config = LoadConfig();
        return config.Currency;
    }

    private AppConfig CreateDefaultConfig()
    {
        return new AppConfig
        {
            DefaultProvider = "openai",
            Providers = new Dictionary<string, ProviderConfig>
            {
                ["openai"] = new ProviderConfig
                {
                    ApiKey = "your-api-key-here",
                    BaseUrl = "https://api.openai.com/v1",
                    Models = new List<string> { "gpt-4", "gpt-3.5-turbo" }
                }
            },
            Proxy = new ProxyConfig
            {
                Host = "127.0.0.1",
                Port = 12306,
                SanitizeRequests = true
            },
            Monitor = new MonitorConfig
            {
                BudgetLimit = 10m,
                WarningThreshold = 75,
                CriticalThreshold = 90,
                DbPath = "contextgate.db"
            },
            Scanner = new ScannerConfig(),
            Context = new ContextConfig(),
            Currency = new CurrencyConfig()
        };
    }
}

File: app/gui-js/lib/core/config-manager.js
const fs = require('fs')
const yaml = require('js-yaml')
const { DEFAULT_PROXY_HOST, DEFAULT_PROXY_PORT } = require('./constants')

class ConfigManager {
  constructor(configPath) {
    this.configPath = configPath
    this.config = this._load()
  }

  _load() {
    try {
      if (fs.existsSync(this.configPath)) {
        return yaml.load(fs.readFileSync(this.configPath, 'utf8')) || {}
      }
    } catch (e) {
      console.error('Failed to load config:', e)
    }
    return {}
  }

  save() {
    try {
      const dir = require('path').dirname(this.configPath)
      if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true })
      fs.writeFileSync(this.configPath, yaml.dump(this.config, { lineWidth: -1 }), 'utf8')
      return true
    } catch (e) {
      console.error('Failed to save config:', e)
      return false
    }
  }

  reload() { this.config = this._load() }

  get(key, defaultValue) {
    const keys = key.split('.')
    let val = this.config
    for (const k of keys) {
      if (val == null) return defaultValue
      val = val[k]
    }
    return val !== undefined ? val : defaultValue
  }

  set(key, value) {
    const keys = key.split('.')
    let obj = this.config
    for (let i = 0; i < keys.length - 1; i++) {
      if (!obj[keys[i]] || typeof obj[keys[i]] !== 'object') obj[keys[i]] = {}
      obj = obj[keys[i]]
    }
    obj[keys[keys.length - 1]] = value
  }

  getWorkspace() { return this.config.workspace || null }
  setWorkspace(p) { this.config.workspace = p; return this.save() }

  getProviders() { return this.config.providers || {} }
  getProvider(name) { return (this.config.providers || {})[name] || null }
  setProvider(name, cfg) {
    if (!this.config.providers) this.config.providers = {}
    this.config.providers[name] = cfg
  }
  removeProvider(name) { if (this.config.providers) delete this.config.providers[name] }

  getDefaultProvider() { return this.config.default_provider || null }
  setDefaultProvider(name) { this.config.default_provider = name }

  getProxyConfig() { return this.config.proxy || { host: DEFAULT_PROXY_HOST, port: DEFAULT_PROXY_PORT } }
  getMonitorConfig() { return this.config.monitor || { budget_limit: 10 } }
  getCurrencyConfig() { return this.config.currency || {} }
  getContextConfig() { return this.config.context || { output_file: 'full_context.txt', max_tokens: 8000, watch_enabled: true, debounce_seconds: 1 } }
  getScannerConfig() { return this.config.scanner || { max_file_size: 1048576 } }
}

module.exports = { ConfigManager }
